nucliadb_vectors/src/data_types/vector.rs

```diff
@@ -41,6 +41,54 @@ pub fn encode_vector(vec: &[Unit]) -> Vec<u8> {
         .fold(encode_length(vec![], vec), encode_unit)
 }
 
+#[inline]
+#[allow(unused)]
+pub fn simd_consine_similarity(mut x: &[u8], mut y: &[u8]) -> Dist {
+    use wide::f32x8;
+    const N: usize = 8;
+    let mut buff_x = [0; 8];
+    let mut buff_y = [0; 8];
+    x.read_exact(&mut buff_x).unwrap();
+    y.read_exact(&mut buff_y).unwrap();
+    let len_x = Len::from_le_bytes(buff_x);
+    let len_y = Len::from_le_bytes(buff_y);
+    assert_eq!(len_x, len_y);
+    let len = len_x;
+    let mut buff_x = [0; 4];
+    let mut buff_y = [0; 4];
+    let mut sum = 0.0;
+    let mut dem_x = 0.0;
+    let mut dem_y = 0.0;
+    let mut i = 0;
+    let mut x_simd_buffer = [0.0; N];
+    let mut y_simd_buffer = [0.0; N];
+    let mut free_in_buffer = 0;
+    while i < len {
+        i += 1;
+        x.read_exact(&mut buff_x).unwrap();
```

Comment:
I'd avoid using read_exact here... The compiler may or may not be smart enough to convert that into a simd unaligned load.

Reply:
Im not sure I follow you on this.. Here im only reading the f32 value, that I have as bytes on x, then I add the f32 to the simd buffer.

Reply:
```
/// `left` & `right` do not need to be aligned.
/// This requires nightly due to portable SIMD,
/// and a cpu with AVX2 or equivalent to run.
/// f32 are assumed to be using native endianess
/// (most likely little endian in your case.)
///
/// AVX 512 should be 2x faster.
pub fn dot_product_simd(left: &[u8], right: &[u8]) -> f32 {
    const DIM: usize = 768;
    const NUM_LANES: usize = 8;
    assert_eq!(left.len(), DIM * 4);
    assert_eq!(right.len(), DIM * 4);
    let mut left_ptr = left.as_ptr() as *const f32x8;
    let mut right_ptr = right.as_ptr() as *const f32x8;
    let mut acc: f32x8 = Default::default();
    unsafe { //< ok thanks to the assert above.
        for i in 0..DIM / NUM_LANES {
            let mut left: f32x8 = std::ptr::read_unaligned(left_ptr);
            let right: f32x8 = std::ptr::read_unaligned(right_ptr);
            acc.add_assign(left.mul(right));
            left_ptr = left_ptr.offset(1 as isize);
            right_ptr = right_ptr.offset(1 as isize);
        }
    }
    acc.as_array().iter().copied().sum()
}
```

Reply:
(You need to add the sqrt root stuff or normalized your vector though)

Reply:
I see, thank you so much! I am trying to adapt this to `nucliadb_vectors` where apart of adding  the sqrt and all that the dimension of the vectors is user defined..

Reply:
for dot product I get a x4 perf improvment with this.

Reply:
here is an implem for cosine...

```rust
pub fn cosine_simd(left: &[u8], right: &[u8]) -> f32 {
    const DIM: usize = 768;
    const NUM_LANES: usize = 8;
    assert_eq!(left.len(), DIM * 4);
    assert_eq!(right.len(), DIM * 4);
    let mut left_ptr = left.as_ptr() as *const f32x8;
    let mut right_ptr = right.as_ptr() as *const f32x8;
    let mut acc_left_norm: f32x8 = Default::default();
    let mut acc_right_norm: f32x8 = Default::default();
    let mut acc_dot_product: f32x8 = Default::default();
    unsafe { //< ok thanks to the assert above.
        for i in 0..DIM / NUM_LANES {
            let mut left: f32x8 = std::ptr::read_unaligned(left_ptr);
            let right: f32x8 = std::ptr::read_unaligned(right_ptr);
            acc_left_norm.add_assign(left.mul(left));
            acc_right_norm.add_assign(right.mul(right));
            acc_dot_product.add_assign(left.mul(right));
            left_ptr = left_ptr.offset(1 as isize);
            right_ptr = right_ptr.offset(1 as isize);
        }
    }
    let norm_left = acc_left_norm.reduce_sum().sqrt();
    let norm_right = acc_right_norm.reduce_sum().sqrt();
    acc_dot_product.reduce_sum() / (norm_left * norm_right)
}
```

Reply:
```
scalar_dot              time:   [564.91 ns 574.43 ns 585.44 ns]
                        change: [-19.913% -16.021% -12.075%] (p = 0.00 < 0.05)
                        Performance has improved.

simd_dot                time:   [74.888 ns 76.182 ns 77.759 ns]
                        change: [+1.3111% +3.2754% +5.5679%] (p = 0.00 < 0.05)
                        Performance has regressed.

scalar_cosine           time:   [601.11 ns 602.38 ns 604.07 ns]
                        change: [+0.3208% +0.7946% +1.2116%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

simd_cosine             time:   [87.091 ns 87.313 ns 87.598 ns]
                        change: [-34.935% -34.503% -34.119%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

```

Reply:
Thank you so much for the help! I am trying to figure out how would this behave in the case where a lane can not be filled, for examen if the dimension of the vectors is 5.

Reply:
I added the following to the code: [398587e](https://github.com/nuclia/nucliadb/pull/575/commits/398587e4f9fd18cc1789112742e2b8a7308bb7d8)
Now I am seeing a better perfomance!

Reply:
can you share the bench? (and make sure you compile with target-cpu=native...)

Reply:
I use [this binary](https://github.com/nuclia/nucliadb/blob/main/vectors_benchmark/src/binaries/1m_stats.rs) for testing the performance.